src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -0,0 +1,3129 @@
+            }
+
+            // Emit the direct parent type
+            writer.WriteLine($"partial {regexClass.Keyword} {regexClass.Name} {regexClass.Constraints}");
+            writer.WriteLine("{");
+            writer.Indent++;
+
+            // Generate a unique name to describe the regex instance.  This includes both
+            // the method name the user provided and a unique counter value, plus a simple
+            // non-randomized (for determinism) hash of the previous content to try to make
+            // the name that much harder to predict.
+            counter++;
+            string generatedName = $"GeneratedRegex_{regexClass.Method.MethodName}_{counter:X}_";
+            generatedName += ComputeStringHash(generatedName).ToString("X");
+
+            // Generate the regex type
+            GenerateRegex(writer, regexClass.Method, generatedName);
+
+            while (writer.Indent != 0)
+            {
+                writer.Indent--;
+                writer.WriteLine("}");
+            }
+
+            // FNV-1a hash function.  The actual algorithm used doesn't matter; just something simple
+            // to create a pseudo-random value based on input text.
+            static uint ComputeStringHash(string s)
+            {
+                uint hashCode = 2166136261;
+                foreach (char c in s)
+                {
+                    hashCode = (c ^ hashCode) * 16777619;
+                }
+                return hashCode;
+            }
+        }
+
+        /// <summary>Gets whether a given regular expression method is supported by the code generator.</summary>
+        private static bool SupportsCustomCodeGeneration(RegexMethod rm)
+        {
+            const RegexOptions SupportedOptions =
+                RegexOptions.IgnoreCase |
+                RegexOptions.Multiline |
+                RegexOptions.ExplicitCapture |
+                RegexOptions.Compiled |
+                RegexOptions.Singleline |
+                RegexOptions.IgnorePatternWhitespace |
+                RegexOptions.RightToLeft |
+                RegexOptions.ECMAScript |
+                RegexOptions.CultureInvariant;
+
+            // If we see an option we're not aware of (but that was allowed through), don't emit custom regex code.
+            return (rm.Options & ~(int)SupportedOptions) == 0;
+        }
+
+        /// <summary>Generates the code for a regular expression method.</summary>
+        private static void GenerateRegex(IndentedTextWriter writer, RegexMethod rm, string id)
+        {
+            string patternExpression = Literal(rm.Pattern);
+            string optionsExpression = $"(RegexOptions)({rm.Options})";
+            string timeoutExpression = rm.MatchTimeout == Timeout.Infinite ?
+                "Timeout.InfiniteTimeSpan" :
+                $"TimeSpan.FromMilliseconds({rm.MatchTimeout.Value.ToString(CultureInfo.InvariantCulture)})";
+
+            writer.WriteLine(s_generatedCodeAttribute);
```

Comment:
nit, it might be easier to read if you wrapped writer.WriteLine eg `wr(s_generatedCodeAttribute)`

in fact where there's a block of lines written together like here, perhaps it would be clearer to have one giant string so there's just one call to write.